emake.mk

```diff
@@ -0,0 +1,101 @@
+# Minimum configuration
+#   PACKAGE_BASENAME: the basename that prefixes all Lisp files in this package
+
+ifndef PACKAGE_BASENAME
+$(error PACKAGE_BASENAME is not set)
+endif
+
+# User options
+#   EMACS_ARGS: [extra arguments for each invocation of emacs]
+#   (below; see README)
+PACKAGE_FILE            ?= $(PACKAGE_BASENAME).el
+PACKAGE_LISP            ?= $(filter-out %-pkg.el %-autoloads.el, $(wildcard $(PACKAGE_BASENAME)*.el))
+PACKAGE_TESTS           ?= $(wildcard test/*.el)
+PACKAGE_ARCHIVES        ?= gnu
+PACKAGE_TEST_ARCHIVES   ?= gnu
+
+# Then, make it easy to invoke Emacs with EMake loaded.
+EMAKE = PACKAGE_FILE="$(PACKAGE_FILE)" \
+	PACKAGE_LISP="$(PACKAGE_LISP)" \
+	PACKAGE_TESTS="$(PACKAGE_TESTS)" \
+	PACKAGE_ARCHIVES="$(PACKAGE_ARCHIVES)" \
+	PACKAGE_TEST_DEPS="$(PACKAGE_TEST_DEPS)" \
+	PACKAGE_TEST_ARCHIVES="$(PACKAGE_TEST_ARCHIVES)" \
+	emacs -batch -l emake.el \
+	--eval "(setq enable-dir-local-variables nil)" \
+	$(EMACS_ARGS) \
+	--eval "(emake (pop argv))"
+
+# Set up our phony targets so Make doesn't think there are files by
+# these names.
+.PHONY: clean setup install compile test help
+
+help:                           ## show help
+	@grep -hE '(^[A-Za-z_/\.\-]+:.*?##.*$$)|(^##.*$$)' $(MAKEFILE_LIST) \
+		| awk 'BEGIN {FS = ":.*?## "}{printf "\033[32m%-30s\033[0m %s\n", $$1, $$2}' \
+		| sed -e 's/\[32m## */[33m/'
+
+
+# Tell Make how to 'clean' this project
+clean:                          ## clean all generated files
+	rm -f *.elc             # delete compiled files
+	rm -rf .elpa/           # delete dependencies
+	rm -rf .elpa.test/
+clean-all: clean                ## clean all generated files including scripts
+	rm -f emacs-travis.mk
+	rm -f emake.el
+
+## Commands useful for Travis
+
+# Tell Make how to 'setup' this project (e.g., for Travis).  This
+# requires both Emacs to be installed and the `emake.el' script to be
+# available.
+setup: emacs emake.el
+
+install: .elpa/                 ## install dependencies
+
+compile: .elpa/                 ## compile the project
+	rm -f *.elc
+	$(EMAKE) compile ~error-on-warn
+
+## Running specific tests
+
+lint-checkdoc: .elpa/           ## checkdoc
+	$(EMAKE) test checkdoc
+
+lint-package-lint: PACKAGE_TEST_DEPS += package-lint
+lint-package-lint: PACKAGE_TEST_ARCHIVES += melpa
+lint-package-lint: .elpa/       ## package-lint
+	$(EMAKE) test package-lint
+
+test-buttercup: PACKAGE_TEST_DEPS += buttercup
+lint-package-lint: PACKAGE_TEST_ARCHIVES += melpa
+test-buttercup: .elpa/          ## buttercup
+	$(EMAKE) test buttercup
+
+test-ert: .elpa/                ## ERT
+	$(EMAKE) test ert
+
+## Support targets
+
+emake.el:                       ## download the EMake script
+	wget 'https://raw.githubusercontent.com/vermiculus/emake.el/master/emake.el'
```

Comment:
if it's ok, could you get `curl` instead of `wget` here? My CI environment doesn't have `wget` by default but I've always seen `curl` everywhere.

Reply:
Absolutely – what would be the appropriate curl invocation?

Reply:
nvm, I just copied your invocation from mpdel 😄

Reply:
sorry, I wanted to write it on the previous message but forgot :-(. This is what [magit uses](https://github.com/magit/magit/blob/188dec9c47e5e63bb52f405f0b81c78693161a28/.travis.yml#L5) except that I replaced short options by long ones for clarity.

```
CURL = curl --fail --silent --show-error --insecure --location --retry 9 --retry-delay 9

emake.el:
	$(CURL) -O https://raw.githubusercontent.com/vermiculus/emake.el/batteries-included/emake.el
```

Reply:
Updated, thanks 😄